Fix #11267: Added audit job to determine the limit of choices length.

SAEb-ai opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #11267.
2. This PR does the following: Adds audit job to determine the limit of choices length.

## Essential Checklist

- [X] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [X] The linter/Karma presubmit checks have passed locally on your machine.
- [X] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [X] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
No proof of changes needed for now.

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

SAEb-ai committed:
Added audit jobs

SAEb-ai committed:
Added test

SAEb-ai committed:
Added to job_registry.py

SAEb-ai committed:
imported ast

SAEb-ai commented:
@aks681 @brianrodri PTAL.

brianrodri commented on core/domain/interaction_jobs_one_off.py:
```diff
@@ -152,6 +152,43 @@ def reduce(key, values):
 
+class MultipleItemInteractionLtOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """Job that produces a list of (exploration ID, length) pairs
+    that use multiple choice interaction or item selection interaction
+    and a list of (exploration ID, (exploration, length)) pairs
+    that use multiple choice interaction or item selection
+    interaction whose choices length is greater than 30.
+    """
```
Please move these to the `map` and `reduce` functions. This doc can just say something like "Audit job that flags MultipleItemInteractions with too many choices".

SAEb-ai replied:
Done.

SAEb-ai committed:
Merge branch 'limit-job2' of github.com:SAEb-ai/oppia into limit-job2

SAEb-ai committed:
Addressed Review Comments

SAEb-ai committed:
Merge remote-tracking branch 'upstream/develop' into limit-job2

SAEb-ai commented:
@brianrodri PTAL.
Thanks for your review comments. I have addressed them.

SAEb-ai commented:
@vojtechjelinek @aks681 PTAL.

brianrodri commented:
Please wait for #12413

vojtechjelinek reviewed (approved):
LGTM!

vojtechjelinek commented:
@SAEb-ai Please fix the backend tests and then feel free to request a run on the backup server.

SAEb-ai committed:
Fix backened Tests Failure

SAEb-ai committed:
Fixed Lint Issues

SAEb-ai committed:
Fixed Lint Issues

SAEb-ai committed:
Fixed Lint Issues

aks681 commented on core/domain/interaction_jobs_one_off.py:
```diff
@@ -152,6 +152,39 @@ def reduce(key, values):
+        exploration = exp_fetchers.get_exploration_from_model(item)
+        interactions_to_check = ('MultipleChoiceInput', 'ItemSelectionInput')
+        for state in exploration.states.items():
+            if state[1].interaction.id in interactions_to_check:
```
Why `state[1]`?

SAEb-ai replied:
`state `is a tuple represented as` (state_name, state_properties)` and thus state_properties being at index 1, I have used `state[1].`

aks681 commented on core/domain/interaction_jobs_one_off.py:
```diff
@@ -152,6 +152,39 @@ def reduce(key, values):
+                for choice in choices.value:
+                    choice_length = len(choice.html)
+                    if choice_length > 30:
+                        yield ('LONGER_THAN_30', (item.id, choice_length))
```
Maybe yield state name as well?

SAEb-ai replied:
The purpose of audit job is [link](https://github.com/oppia/oppia/issues/11267#issuecomment-783259128) and I don't think so yeilding state_name is required in this job.

aks681 reviewed (commented):
Couple of minor comments, else lgtm as codeowner!

sajalasati commented:
@SAEb-ai Any updates on this PR?

SAEb-ai committed:
Merge remote-tracking branch 'upstream/develop' into limit-job2

SAEb-ai commented:
@aks681 Thanks for your review comments.I have left some comments from my side too PTAL.
Thanks!

SAEb-ai commented:
> @SAEb-ai Any updates on this PR?

@sajalasati I have filled the form for this job to be run on the backup server.

SAEb-ai commented:
@sajalasati Can you please mail @aks681 regarding the review of this PR as oppiabot has now added the stale label to this PR now.

@aks681 There is some more discussion required  for determining the choice length as vojta has ran this job on the backup server but there were some problems coming regarding the choice length for which vojta has added you and kevin to the email thread. PTAL at the email thread.

@kevintab95 I am also assigning you to this PR as vojta has informed me to discuss the choices length with you too and you are also added to the email thread of the results of this job run on the backup server. PTAL at that too.

Thanks!

kevintab95 commented:
Replied to the email. Thanks!

SAEb-ai commented:
@kevintab95 Assigning you again so that you can again reply to the email as oppiabot has added the stale label again. Thanks!

@DubeySandeep Can you please ask @aks681 to review this PR  as aks681 hasn't provided his review comments from about a month. Thanks!

DubeySandeep commented:
@SAEb-ai Sorry for the trouble!

@aks681 Can you please review this PR?

aks681 reviewed (approved):
Lgtm!

vojtechjelinek commented:
@SAEb-ai Does this need to be merged? The job was run on the backup server and I don't see why would we need this on the production?

SAEb-ai commented:
@vojtechjelinek I think @kevintab95 can better answer your question regarding this as he had only suggested https://github.com/oppia/oppia/issues/11267#issuecomment-781601694

kevintab95 commented:
Hi, I agree that this need not be merged. We already have the data regarding the length of choices. We just need to decide what the ideal max length should be looking at the data. We can prioritize this as part the validation project in the LaCE team. 

/cc @EricZLou FYI
